packages/client-sdk-nodejs/src/internal/leaderboard-client.ts

```diff
@@ -0,0 +1,548 @@
+import {
+  CredentialProvider,
+  InvalidArgumentError,
+  LeaderboardDelete,
+  LeaderboardFetch,
+  LeaderboardLength,
+  LeaderboardRemoveElements,
+  LeaderboardUpsert,
+  MomentoLogger,
+  MomentoLoggerFactory,
+  LeaderboardOrder,
+} from '@gomomento/sdk-core';
+import {LeaderboardClientProps} from '../leaderboard-client-props';
+import {normalizeSdkError} from '@gomomento/sdk-core/dist/src/errors';
+import {
+  validateCacheName,
+  validateLeaderboardName,
+  validateLeaderboardNumberOfElements,
+  validateSortedSetScores,
+  validateLeaderboardOffset,
+  validateLeaderboardCount,
+  validateLeaderboardRanks,
+} from '@gomomento/sdk-core/dist/src/internal/utils';
+import {LeaderboardConfiguration} from '../config/leaderboard-configuration';
+import {
+  InternalLeaderboardClient,
+  LeaderboardFetchByRankCallOptions,
+  LeaderboardFetchByScoreCallOptions,
+  LeaderboardGetRankCallOptions,
+} from '@gomomento/sdk-core/dist/src/internal/clients';
+import {leaderboard} from '@gomomento/generated-types/dist/leaderboard';
+import _Element = leaderboard._Element;
+import {IdleGrpcClientWrapper} from './grpc/idle-grpc-client-wrapper';
+import {GrpcClientWrapper} from './grpc/grpc-client-wrapper';
+import {Header, HeaderInterceptorProvider} from './grpc/headers-interceptor';
+import {ClientTimeoutInterceptor} from './grpc/client-timeout-interceptor';
+import {cacheServiceErrorMapper} from '../errors/cache-service-error-mapper';
+import {
+  ChannelCredentials,
+  Interceptor,
+  Metadata,
+  ServiceError,
+  status,
+} from '@grpc/grpc-js';
+import {version} from '../../package.json';
+import {_RankedElement} from '@gomomento/sdk-core/dist/src/messages/responses/grpc-response-types';
+
+export class LeaderboardDataClient implements InternalLeaderboardClient {
+  public readonly cacheName: string;
+  public readonly leaderboardName: string;
+  private readonly configuration: LeaderboardConfiguration;
+  private readonly credentialProvider: CredentialProvider;
+  private readonly logger: MomentoLogger;
+  private readonly requestTimeoutMs: number;
+  private readonly clientWrapper: GrpcClientWrapper<leaderboard.LeaderboardClient>;
+  private readonly interceptors: Interceptor[];
+
+  constructor(
+    props: LeaderboardClientProps,
+    cacheName: string,
+    leaderboardName: string
+  ) {
+    try {
+      validateCacheName(cacheName);
+      validateLeaderboardName(leaderboardName);
+    } catch (err) {
+      throw new InvalidArgumentError(
+        'cache name and leaderboard name must not be empty strings'
+      );
+    }
```

Comment:
Is it reasonable to validate the cache and leaderboard names in the constructor and throw an error? Or does that make for an unexpected UX?

Reply:
I think that's fine.